pkg/common/cns-lib/volume/manager.go

```diff
@@ -1014,3 +1016,61 @@ func (m *defaultManager) RetrieveVStorageObject(ctx context.Context, volumeID st
 	log.Debugf("vStorageObject for volumeID: %q is %+v", volumeID, vStorageObject)
 	return vStorageObject, nil
 }
+
+// QueryVolumeAsync returns volumes matching the given filter using CnsQueryAsync API.
+func (m *defaultManager) QueryVolumeAsync(ctx context.Context, queryFilter cnstypes.CnsQueryFilter, querySelection cnstypes.CnsQuerySelection) (*cnstypes.CnsQueryResult, error) {
+	internalQueryVolumeAsync := func() (*cnstypes.CnsQueryResult, error) {
+		log := logger.GetLogger(ctx)
+		err := validateManager(ctx, m)
+		if err != nil {
+			return nil, err
+		}
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		// Set up the VC connection
+		err = m.virtualCenter.ConnectCns(ctx)
+		if err != nil {
+			log.Errorf("ConnectCns failed with err: %+v", err)
+			return nil, err
+		}
+		//Call the CNS QueryVolumeAsync
+		queryVolumeAsyncTask, err := m.virtualCenter.CnsClient.QueryVolumeAsync(ctx, queryFilter, querySelection)
+		if err != nil {
+			log.Errorf("CNS QueryVolumeAsync failed from vCenter %q with err: %v", m.virtualCenter.Config.Host, err)
+			return nil, err
+		}
+		queryVolumeAsyncTaskInfo, err := cns.GetTaskInfo(ctx, queryVolumeAsyncTask)
+		if err != nil {
+			log.Errorf("CNS QueryVolumeAsync failed from vCenter %q with err: %v", m.virtualCenter.Config.Host, err)
+			return nil, err
+		}
+		queryVolumeAsyncTaskResult, err := cns.GetTaskResult(ctx, queryVolumeAsyncTaskInfo)
+		if err != nil {
+			log.Errorf("CNS QueryVolumeAsync failed from vCenter %q with err: %v", m.virtualCenter.Config.Host, err)
+			return nil, err
+		}
+		if queryVolumeAsyncTaskResult == nil {
+			log.Errorf("queryVolumeAsyncTaskResult is empty for QueryVolumeAsync task: %q, opID: %q", queryVolumeAsyncTaskInfo.Task.Value, queryVolumeAsyncTaskInfo.ActivationId)
+			return nil, errors.New("taskResult is empty")
+		}
+		volumeOperationRes := queryVolumeAsyncTaskResult.GetCnsVolumeOperationResult()
+		if volumeOperationRes.Fault != nil {
+			msg := fmt.Sprintf("failed to query volumes using CnsQueryVolumeAsync, fault: %q, opID: %q", spew.Sdump(volumeOperationRes.Fault), queryVolumeAsyncTaskInfo.ActivationId)
+			log.Error(msg)
+			return nil, errors.New(msg)
+		}
+		queryVolumeAsyncResult := interface{}(queryVolumeAsyncTaskResult).(*cnstypes.CnsQueryResult)
+		log.Infof("QueryVolumeAsync successfully returned CnsQueryResult: %+v, opId: %q", spew.Sdump(queryVolumeAsyncResult), queryVolumeAsyncTaskInfo.ActivationId)
```

Comment:
Do we want `spec.Sdump()` as part of info level logs?

Reply:
Added a debug log to capture spec in logs